
# 🧮 NumPy Mega Homework

This workbook contains **lots** of NumPy practice tasks, grouped by topic.  
For each task, write your solution in the empty code cell that follows it.

**Rules**
- Use **NumPy** (`import numpy as np`) for array operations.
- Do **not** use Python lists for computations unless explicitly asked.
- Add comments to explain key steps.

**Sections**
1. Array Basics & Creation (10 tasks)  
2. Indexing & Slicing (8 tasks)  
3. Array Operations & Broadcasting (8 tasks)  
4. Mathematical Functions (6 tasks)  
5. Aggregations & Descriptive Stats (6 tasks)  
6. Random Numbers & Simulations (8 tasks)  
7. Linear Algebra (6 tasks)  
8. Reshaping, Stacking & Splitting (5 tasks)  
9. Real‑World Mini Case Studies (7 tasks)


## 1) Array Basics & Creation

**1.1** Import NumPy as `np` and print the installed version.

In [2]:
import numpy as np
print(np.__version__)

2.5.0


**1.2** Create a 1D array with values from 0 to 20 (inclusive).

In [3]:
import numpy as np
array=np.arange(21)
print(array)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20]


**1.3** Create a 1D array of the first 15 even numbers.

In [5]:
import numpy as np
evens=2*np.arange(1,16)
print(evens)

[ 2  4  6  8 10 12 14 16 18 20 22 24 26 28 30]


**1.4** Create a 3×3 array filled with 7s (use `np.full`).

In [6]:
import numpy as np
array=np.full((3,3),7)
print(array)

[[7 7 7]
 [7 7 7]
 [7 7 7]]


**1.5** Create a 5×5 identity matrix.

In [8]:
import numpy as np
İ=np.identity(5)
print(İ)

[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]


**1.6** Create an array of 10 numbers linearly spaced between 0 and 1 (inclusive).

In [9]:
import numpy as np
array=np.linspace(0,1,10)
print(array)

[0.         0.11111111 0.22222222 0.33333333 0.44444444 0.55555556
 0.66666667 0.77777778 0.88888889 1.        ]


**1.7** Create an array of 12 values logarithmically spaced between 10^0 and 10^3.

In [10]:
import numpy as np
array=np.logspace(0,3,12)
print(array)

[   1.            1.87381742    3.51119173    6.57933225   12.32846739
   23.101297     43.28761281   81.11308308  151.9911083   284.80358684
  533.66992312 1000.        ]


**1.8** Create a boolean array of length 12 that alternates `True, False, True, False, ...`.

In [11]:
import numpy as np
array = (np.arange(12) % 2) == 0  # True, False, True, ...
print(array)

[ True False  True False  True False  True False  True False  True False]


**1.9** Create a 4×4 array of random integers from 10 to 99 (inclusive).

In [13]:
import numpy as np
np.random.seed(0)
array = np.random.randint(10, 100, size=(4, 4))
print(array)

[[54 57 74 77]
 [77 19 93 31]
 [46 97 80 98]
 [98 22 68 75]]


**1.10** Given `a = np.array([1, 2, 3], dtype=np.int32)`, change its dtype to float64 without copying if possible.

In [15]:
import numpy as np
a=np.array([1,2,3], dtype=np.int32)
b=a.astype(np.float64)
print(b,b.dtype)
print(np.shares_memory(a,b))

[1. 2. 3.] float64
False


## 2) Indexing & Slicing

**2.1** Create an array `x = np.arange(1, 26).reshape(5,5)`. Extract the **center 3×3 block**.

In [16]:
import numpy as np
x = np.arange(1, 26).reshape(5, 5)
center = x[1:4, 1:4]
print(center)

[[ 7  8  9]
 [12 13 14]
 [17 18 19]]


**2.2** From `x` above, extract the **border elements** (top row, bottom row, leftmost column, rightmost column) into a 1D array.

In [17]:
import numpy as np
x = np.arange(1, 26).reshape(5, 5)
top = x[0, :]
bottom = x[-1, :]
left = x[:, 0]
right = x[:, -1]
border = np.concatenate([top, bottom, left, right])
print(border)

[ 1  2  3  4  5 21 22 23 24 25  1  6 11 16 21  5 10 15 20 25]


**2.3** From `x`, select every other row and every other column.

In [18]:
import numpy as np
x = np.arange(1, 26).reshape(5, 5)
sel = x[::2, ::2]
print(sel)

[[ 1  3  5]
 [11 13 15]
 [21 23 25]]


**2.4** Given `y = np.arange(10)`, reverse it using slicing only.

In [19]:
import numpy as np
y = np.arange(10)
rev = y[::-1]
print(rev)

[9 8 7 6 5 4 3 2 1 0]


**2.5** From `y`, extract elements with **odd indices** (1,3,5,...) using slicing.

In [20]:
import numpy as np
y = np.arange(10)
odd_indices = y[1::2]
print(odd_indices)

[1 3 5 7 9]


**2.6** Use **fancy indexing** to pick elements at indices `[0, 3, 7]` from `y`.

In [21]:
import numpy as np
y = np.arange(10)
picked = y[[0, 3, 7]]
print(picked)

[0 3 7]


**2.7** Given `z = np.array([3, -1, 7, 0, -5, 2])`, replace **negative** values with `0` using boolean indexing.

In [22]:
import numpy as np
z = np.array([3, -1, 7, 0, -5, 2])
z[z < 0] =0
print(z)

[3 0 7 0 0 2]


**2.8** Given `M = np.arange(1,17).reshape(4,4)`, set the **diagonal** to `-1` without a loop.

In [23]:
import numpy as np
M = np.arange(1, 17).reshape(4, 4)
np.fill_diagonal(M, -1)
print(M)

[[-1  2  3  4]
 [ 5 -1  7  8]
 [ 9 10 -1 12]
 [13 14 15 -1]]


## 3) Array Operations & Broadcasting

**3.1** Compute element‑wise addition, subtraction, multiplication, and division of arrays `a=[1,2,3,4]` and `b=[10,20,30,40]`.

In [24]:
import numpy as np
a = np.array([1,2,3,4])
b = np.array([10,20,30,40])
add = a + b
sub = a - b
mul = a * b
div = a / b
print('add:', add)
print('sub:', sub)
print('mul:', mul)
print('div:', div)

add: [11 22 33 44]
sub: [ -9 -18 -27 -36]
mul: [ 10  40  90 160]
div: [0.1 0.1 0.1 0.1]


**3.2** Square all elements of `a` without using a Python loop.

In [1]:
import numpy as np
a = np.array([1, 2, 3, 4])
squared = a ** 2
print('squared (new):', squared)
a_inplace = np.array([1, 2, 3, 4])
np.square(a_inplace, out=a_inplace)
print('squared (in-place):', a_inplace)

squared (new): [ 1  4  9 16]
squared (in-place): [ 1  4  9 16]


**3.3** Let `A = np.ones((3,3))` and `v = np.array([1,2,3])`. Add `v` to each **row** of `A` using broadcasting.

In [1]:
import numpy as np
A=np.ones((3,3))
v=np.array([1,2,3])
result=A+v
print(result)

[[2. 3. 4.]
 [2. 3. 4.]
 [2. 3. 4.]]


**3.4** Multiply each **column** of `A` by `[1,2,3]` using broadcasting.

In [5]:
import numpy as np
A=np.ones((3,3))
col_vector=np.array([1,2,3])
result=A*col_vector
print(result)

[[1. 2. 3.]
 [1. 2. 3.]
 [1. 2. 3.]]


**3.5** Normalize array `p = np.array([3.0, 4.0, 0.0, 8.0])` to unit length (divide by its Euclidean norm).

In [7]:
import numpy as np
p=np.array([3.0,4.0,5.0])
p_norm=np.linalg.norm(p)
normalized=p/p_norm
print('norm:',p_norm)
print('normalized:',normalized)

norm: 7.0710678118654755
normalized: [0.42426407 0.56568542 0.70710678]


**3.6** Compute the pairwise sum of vectors in `X = np.array([[1,2],[3,4],[5,6]])` with `y = np.array([10,20])` via broadcasting (result shape 3×2).

In [10]:
import numpy as np
X=np.array([[1,2,],[3,4],[5,6]])
Y=np.array([10,20])
result=X+Y
print(result)

[[11 22]
 [13 24]
 [15 26]]


**3.7** Given `B = np.arange(12).reshape(3,4)`, subtract the **column means** from each column (column‑wise centering) using broadcasting.

In [11]:
import numpy as np
B=np.arange(12).reshape(3,4)
col_means=B.mean(axis=0)
centered=B-col_means
print(centered)

[[-4. -4. -4. -4.]
 [ 0.  0.  0.  0.]
 [ 4.  4.  4.  4.]]


**3.8** Clip array values: given `w = np.array([-3, -1, 0, 2, 9])`, clip to the range `[0, 5]`.

In [12]:
import numpy as np
w=np.array([-3,-1,0,2,9])
clipped=np.clip(w,0,5)
print(clipped)

[0 0 0 2 5]


## 4) Mathematical Functions

**4.1** For `q = np.linspace(-2*np.pi, 2*np.pi, 50)`, compute `sin(q)`, `cos(q)`, and `tan(q)` arrays.

In [15]:
import numpy as np
q=np.linspace(-2*np.pi,2*np.pi,50)
sin_q=np.sin(q)
cos_q=np.cos(q)
tan_q=np.tan(q)
print('sin(q):',sin_q)
print('cos(q):',cos_q)
print('tan(q):',tan_q)


sin(q): [ 2.44929360e-16  2.53654584e-01  4.90717552e-01  6.95682551e-01
  8.55142763e-01  9.58667853e-01  9.99486216e-01  9.74927912e-01
  8.86599306e-01  7.40277997e-01  5.45534901e-01  3.15108218e-01
  6.40702200e-02 -1.91158629e-01 -4.33883739e-01 -6.48228395e-01
 -8.20172255e-01 -9.38468422e-01 -9.95379113e-01 -9.87181783e-01
 -9.14412623e-01 -7.81831482e-01 -5.98110530e-01 -3.75267005e-01
 -1.27877162e-01  1.27877162e-01  3.75267005e-01  5.98110530e-01
  7.81831482e-01  9.14412623e-01  9.87181783e-01  9.95379113e-01
  9.38468422e-01  8.20172255e-01  6.48228395e-01  4.33883739e-01
  1.91158629e-01 -6.40702200e-02 -3.15108218e-01 -5.45534901e-01
 -7.40277997e-01 -8.86599306e-01 -9.74927912e-01 -9.99486216e-01
 -9.58667853e-01 -8.55142763e-01 -6.95682551e-01 -4.90717552e-01
 -2.53654584e-01 -2.44929360e-16]
cos(q): [ 1.          0.96729486  0.8713187   0.71834935  0.51839257  0.28452759
  0.03205158 -0.22252093 -0.46253829 -0.67230089 -0.8380881  -0.94905575
 -0.99794539 -0.98155916

**4.2** Compute element‑wise `exp`, `log`, and `sqrt` for `r = np.array([1, e, e**2, 10.0])` (use `np.e` where needed).

In [16]:
import numpy as np
r=np.array([1,np.e,np.e**2, 10.0])
exp_r=np.exp(r)
log_r=np.log(r)
sqrt_r=np.sqrt(r)
print('exp(r):',exp_r)
print('log(r):',log_r)
print('sqrt(r):',sqrt_r)

exp(r): [2.71828183e+00 1.51542622e+01 1.61817799e+03 2.20264658e+04]
log(r): [0.         1.         2.         2.30258509]
sqrt(r): [1.         1.64872127 2.71828183 3.16227766]


**4.3** Compute `np.floor`, `np.ceil`, and `np.round` for `s = np.array([1.2, 3.7, -2.3, -4.8])`.

In [17]:
import numpy as np
s=np.array([1.2,3.7,-2.3,-4.8])
floor_s=np.floor(s)
ceil_s=np.ceil(s)
round_s=np.round(s)
print('floor:',floor_s)
print('ceil:',ceil_s)
print('round:',round_s)


floor: [ 1.  3. -3. -5.]
ceil: [ 2.  4. -2. -4.]
round: [ 1.  4. -2. -5.]


**4.4** Safely compute `log1p(x)` for `x = np.array([0, 1e-10, 1e-5, 0.1])`. Why is `log1p` preferred here? (Answer in a comment)

In [18]:
import numpy as np
x=np.array([0,1e-10,1e-5,0.1])
result=np.log1p(x)
print(result)



[0.00000000e+00 1.00000000e-10 9.99995000e-06 9.53101798e-02]


**4.5** Compute the sigmoid function `σ(x) = 1 / (1 + exp(-x))` for `x = np.linspace(-8, 8, 9)`.

In [19]:
import numpy as np
x=np.linspace(-8, 8, 9)
sigmoid=1/(1+np.exp(-x))
print(sigmoid)

[3.35350130e-04 2.47262316e-03 1.79862100e-02 1.19202922e-01
 5.00000000e-01 8.80797078e-01 9.82013790e-01 9.97527377e-01
 9.99664650e-01]


**4.6** Standardize `t` (z‑score): given `t = np.array([10, 12, 9, 13, 11, 8])`, transform to mean 0 and std 1.

In [23]:
import numpy as np
t=np.array([10, 12, 9, 13, 11, 8])
mean_t=t.mean()
std_t=t.std()
z_score=(t-mean_t)/std_t
print(z_score)

[-0.29277002  0.87831007 -0.87831007  1.46385011  0.29277002 -1.46385011]


## 5) Aggregations & Descriptive Stats

**5.1** Given `u = np.array([5, 5, 9, 1, 1, 9, 9])`, compute: mean, median, variance, std, min, max.

In [24]:
import numpy as np
u=np.array([5, 5, 9 , 1, 1, 9, 9])
mean_u=np.mean(u)
median_u=np.median(u)
var_u=np.var(u)
std_u=np.std(u)
min_u=np.min(u)
max_u=np.max(u)
print('mean:', mean_u)
print('median:', median_u)
print('variance:', var_u)
print('standard deviation:', std_u)
print('minimum:', min_u)
print('maximum:', max_u)

mean: 5.571428571428571
median: 5.0
variance: 11.10204081632653
standard deviation: 3.3319725113401715
minimum: 1
maximum: 9


**5.2** Compute `argmin` and `argmax` for `u`. What values do they point to? (Print index & value)

In [25]:
import numpy as np
u=np.array([5, 5, 9, 1, 1, 9, 9])
argmin_u=np.argmin(u)
argmax_u=np.argmax(u)
print('argmin:', argmin_u, 'value:', u[argmin_u])
print('argmax:', argmax_u, 'value:', u[argmax_u])

argmin: 3 value: 1
argmax: 2 value: 9


**5.3** For `V = np.arange(12).reshape(3,4)`, compute **row sums** and **column means**.

In [26]:
import numpy as np
V=np.arange(12).reshape(3,4)
row_sums=V.sum(axis=1)
col_means=V.mean(axis=0)
print('row sums:', row_sums)
print('column means:', col_means)

row sums: [ 6 22 38]
column means: [4. 5. 6. 7.]


**5.4** For `V`, compute the cumulative sum along axis 1 and the cumulative product along axis 0.

In [27]:
import numpy as np
V=np.arange(12).reshape(3,4)
cumulative_sum_rows=np.cumsum(V, axis=1)
cumulative_product_cols=np.cumprod(V, axis=0)
print('cumulative sum along rows:')
print(cumulative_sum_rows)
print('cumulative product along columns:')
print(cumulative_product_cols)

cumulative sum along rows:
[[ 0  1  3  6]
 [ 4  9 15 22]
 [ 8 17 27 38]]
cumulative product along columns:
[[  0   1   2   3]
 [  0   5  12  21]
 [  0  45 120 231]]


**5.5** For `u`, return the **unique** values and their **counts** (hint: `np.unique(..., return_counts=True)`).

In [28]:
import numpy as np
u=np.array([5, 5, 9, 1, 1, 9, 9])
unique_values, counts=np.unique(u, return_counts=True)
print('unique values:', unique_values)
print('counts:', counts)

unique values: [1 5 9]
counts: [2 2 3]


**5.6** Compute the **50th percentile (median)** and **90th percentile** of `u` using `np.percentile`.

In [29]:
import numpy as np
u=np.array([5, 5, 9, 1, 1, 9, 9])
p50=np.percentile(u, 50)
p90=np.percentile(u, 90)
print('50th percentile:', p50)
print('90th percentile:', p90)

50th percentile: 5.0
90th percentile: 9.0


## 6) Random Numbers & Simulations

**6.1** Set a random seed (e.g., 123). Generate 10 random numbers from the uniform distribution [0,1).

In [30]:
import numpy as np
np.random.seed(123)
random_numbers=np.random.rand(10)
print(random_numbers)

[0.69646919 0.28613933 0.22685145 0.55131477 0.71946897 0.42310646
 0.9807642  0.68482974 0.4809319  0.39211752]


**6.2** Generate a 3×3 matrix of random integers between 1 and 100 (inclusive).

In [31]:
import numpy as np
matrix=np.random.randint(1, 100, size=(3,3))
print(matrix)

[[47 97 26]
 [84 79 37]
 [97 81 69]]


**6.3** Draw 1,000 samples from a standard normal distribution and compute sample mean and std. Compare to theoretical values.

In [32]:
import numpy as np
samples=np.random.randn(1000)
sample_mean=samples.mean()
sample_std=samples.std()
print('sample mean:', sample_mean)
print('sample std:', sample_std)
print('sample std:', sample_std)
print('theoretical mean:', 0)
print('theoretical std:', 1)

sample mean: 0.01689973953857308
sample std: 1.059548563986532
sample std: 1.059548563986532
theoretical mean: 0
theoretical std: 1


**6.4** Simulate 1,000 fair coin flips (use 0/1). Estimate the probability of heads.

In [33]:
import numpy as np
flips=np.random.randint(0, 2, size=1000)
prob_heads=flips.mean()
print('probability of heads:', prob_heads)

probability of heads: 0.482


**6.5** Simulate rolling two fair dice 10,000 times. Estimate the probability that the sum equals 7.

In [34]:
import numpy as np
dice1=np.random.randint(1,7,size=1000)
dice2=np.random.randint(1,7,size=1000)
sums=dice1+dice2
prob_sums_7=np.mean(sums==7)
print('probability of sum being 7:', prob_sums_7)

probability of sum being 7: 0.16


**6.6** Draw 500 samples from a binomial distribution (n=20, p=0.4). Compute mean and variance; compare to theory.

In [35]:
import numpy as np
samples=np.random.binomial(n=20, p=0.4, size=500)
sample_mean=samples.mean()
sample_var=samples.var()
print('sample mean:', sample_mean)
print('sample variance:', sample_var)
print('theoretical mean:', 20 * 0.4)
print('theoretical variance:', 20 * 0.4 * (1 - 0.4))

sample mean: 7.988
sample variance: 4.779856
theoretical mean: 8.0
theoretical variance: 4.8


**6.7** Use `np.random.multivariate_normal` to draw 1,000 2D samples with mean `[0,0]` and covariance `[[1,0.5],[0.5,1]]`. Compute the sample covariance matrix.

In [39]:
import numpy as np
mean=[0, 0]
cov=[[1, 0.5], [0.5, 1]]
samples=np.random.multivariate_normal(mean, cov, size=500)
sample_cov=np.cov(samples, rowvar=False)
print(sample_cov)

[[1.03724613 0.47712955]
 [0.47712955 0.95689453]]


**6.8** Monte Carlo: Estimate π by randomly sampling 100,000 points in the unit square and counting how many fall inside the unit quarter‑circle.

In [40]:
import numpy as np
rng=np.random.default_rng(42)
x=rng.random(100000)
y=rng.random(100000)
inside=x*x+y*y <= 1
pi_estimate=4*inside.mean()
print('estimated pi:', pi_estimate)

estimated pi: 3.13776


## 7) Linear Algebra

**7.1** Compute the dot product of `a = [1,2,3]` and `b = [4,5,6]`.

In [41]:
import numpy as np
a=np.array([1,2,3])
b=np.array([4,5,6])
dot_product=np.dot(a,b)
print(dot_product)

32


**6.8** Monte Carlo: Estimate π by randomly sampling 100,000 points in the unit square and counting how many fall inside the unit quarter‑circle.

**7.2** Given `A = [[2,1],[1,2]]`, compute determinant, inverse, and verify `A @ A_inv ≈ I`.

In [42]:
import numpy as np
A=np.array([[2, 1], [1, 2]])
det_A=np.linalg.det(A)
A_inv=np.linalg.inv(A)
product=A@A_inv
print('determinant:', det_A)
print('inverse:\n',A_inv)
print('A@A_inv:\n',product)
print('A@A_inv is close to identity:', np.allclose(product, np.eye(2)))

determinant: 2.9999999999999996
inverse:
 [[ 0.66666667 -0.33333333]
 [-0.33333333  0.66666667]]
A@A_inv:
 [[1. 0.]
 [0. 1.]]
A@A_inv is close to identity: True


**7.3** For `A` above, compute eigenvalues and eigenvectors, and verify `A @ v ≈ λ v`.

In [43]:
import numpy as np
A=np.array([[2, 1], [1, 2]])
eigenvalues,eigenvectors=np.linalg.eig(A)
print('eigenvalues:', eigenvalues)
print('eigenvectors:\n', eigenvectors)
for i, lam in enumerate(eigenvalues):
    v=eigenvectors[:, i]
    print(f'chek for eigenvalue{lam}:', np.allclose(A @ v, lam * v))


eigenvalues: [3.+0.j 1.+0.j]
eigenvectors:
 [[ 0.70710678+0.j -0.70710678+0.j]
 [ 0.70710678+0.j  0.70710678+0.j]]
chek for eigenvalue(3+0j): True
chek for eigenvalue(1+0j): True


**7.4** Solve the linear system `3x + 2y = 7` and `2x + 5y = 8`.

In [44]:
import numpy as np
A=np.array([[3, 2], [2,5]])
b=np.array([7, 8])
x=np.linalg.solve(A,b)
print(x)

[1.72727273 0.90909091]


**7.5** Create a random 4×4 matrix and compute its **rank**.

In [45]:
import numpy as np
rng=np.random.default_rng(42)
M=rng.integers(0, 10, size=(4, 4))
rank_M=np.linalg.matrix_rank(M)
print(M)
print('rank:', rank_M)

[[0 7 6 4]
 [4 8 0 6]
 [2 0 5 9]
 [7 7 7 7]]
rank: 4


**7.6** Perform QR decomposition of a random 4×3 matrix and verify `Q @ R ≈ A`.

In [46]:
import numpy as np
rng = np.random.default_rng(42)
A = rng.normal(size=(4, 3))
Q, R = np.linalg.qr(A)
reconstructed = Q @ R
print('Q shape:', Q.shape)
print('R shape:', R.shape)
print('Q:\n', Q)
print('R:\n', R)
print('reconstructed A:\n', reconstructed)
print('is close to original?', np.allclose(reconstructed, A))

Q shape: (4, 3)
R shape: (3, 3)
Q:
 [[-0.23224054 -0.60685107 -0.74833931]
 [-0.7168527  -0.40137244  0.56344296]
 [-0.09743374 -0.11424096 -0.05305216]
 [ 0.65014861 -0.67644729  0.34598522]]
R:
 [[-1.3120753   2.24268337  1.26650304]
 [ 0.          0.85546871 -0.45696901]
 [ 0.          0.         -1.02530016]]
reconstructed A:
 [[ 0.30471708 -1.03998411  0.7504512 ]
 [ 0.94056472 -1.95103519 -1.30217951]
 [ 0.1278404  -0.31624259 -0.01680116]
 [-0.85304393  0.87939797  0.77779194]]
is close to original? True


## 8) Reshaping, Stacking & Splitting

**8.1** Create `a = np.arange(24)` and reshape to (4,6).

In [47]:
import numpy as np
a = np.arange(24).reshape(4, 6)
print(a)

[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]


**8.2** From the (4,6) array, extract a (2,3) block from the center.

In [48]:
import numpy as np
a = np.arange(24).reshape(4, 6)
block = a[1:3, 1:4]
print(block)

[[ 7  8  9]
 [13 14 15]]


**8.3** Vertically stack two arrays of shape (2,3) to obtain (4,3); then horizontally stack to obtain (2,6).

In [49]:
import numpy as np

x = np.array([[1, 2, 3], [4, 5, 6]])
y = np.array([[7, 8, 9], [10, 11, 12]])

vertical = np.vstack((x, y))
horizontal = np.hstack((x, y))

print('Vertical stack:\n', vertical)
print('Horizontal stack:\n', horizontal)

Vertical stack:
 [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
Horizontal stack:
 [[ 1  2  3  7  8  9]
 [ 4  5  6 10 11 12]]


**8.4** Split an array of length 20 into 4 equal parts; then split into parts of lengths `[3,5,12]`.

In [2]:
import numpy as np
array=np.arange(20)
equal_parts=np.array_split(array,4)
custom_parts=np.array_split(array,[3, 8])
print('split into 4 equal parts:')
for i, part in enumerate(equal_parts, 1):
    print(i, part)
print('\nsplit into length [3, 5, 12]')
for i, part in enumerate(custom_parts, 1):
    print(i, part)

split into 4 equal parts:
1 [0 1 2 3 4]
2 [5 6 7 8 9]
3 [10 11 12 13 14]
4 [15 16 17 18 19]

split into length [3, 5, 12]
1 [0 1 2]
2 [3 4 5 6 7]
3 [ 8  9 10 11 12 13 14 15 16 17 18 19]


**8.5** Use `np.concatenate` to combine three arrays along axis 0 and along axis 1 (where shapes allow).

In [1]:
import numpy as np

arr1 = np.array([[1, 2], [3, 4]])
arr2 = np.array([[5, 6], [7, 8]])
arr3 = np.array([[9, 10], [11, 12]])

concat_axis0 = np.concatenate((arr1, arr2, arr3), axis=0)
concat_axis1 = np.concatenate((arr1, arr2, arr3), axis=1)

print('Along axis 0:\n', concat_axis0)
print('\nAlong axis 1:\n', concat_axis1)

Along axis 0:
 [[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]
 [11 12]]

Along axis 1:
 [[ 1  2  5  6  9 10]
 [ 3  4  7  8 11 12]]


## 9) Real‑World Mini Case Studies

**9.1 Finance**: Prices of a stock for 10 days are `p = [100, 102, 101, 103, 105, 104, 106, 107, 110, 108]`. Compute **daily returns** (percentage) and **cumulative return**.

In [2]:
import numpy as np

p = np.array([100, 102, 101, 103, 105, 104, 106, 107, 110, 108])

returns = (p[1:] / p[:-1] - 1) * 100
cumulative_return = (p[-1] / p[0] - 1) * 100

print('Daily returns (%):', returns)
print('Cumulative return (%):', cumulative_return)

Daily returns (%): [ 2.         -0.98039216  1.98019802  1.94174757 -0.95238095  1.92307692
  0.94339623  2.80373832 -1.81818182]
Cumulative return (%): 8.000000000000007


**9.2 Demography**: Ages of 20 people are in an array. Compute mean, median, std, and the proportion aged 18–30. *(Generate random integers 10–80 with a fixed seed.)*

In [ ]:
import numpy as np
np.random.seed(42)
ages = np.random.randint(10, 81, size=20)
mean_age = ages.mean()
median_age = np.median(ages)
std_age = ages.std()
proportion_18_30 = np.mean((ages >= 18) & (ages <= 30))
print('Ages:', ages)
print('Mean:', mean_age)
print('Median:', median_age)
print('Std:', std_age)
print('Proportion aged 18-30:', proportion_18_30)

**9.3 Quality Control**: Simulate 1,000 measurements with true value 50 and measurement noise `N(0, 2^2)`. What fraction of measurements fall within ±3 of the true value?

In [3]:
import numpy as np
np.random.seed(0)
true_value = 50
noise = np.random.normal(loc=0, scale=2, size=1000)
measurements = true_value + noise
within_3 = np.mean(np.abs(measurements - true_value) <= 3)
print('Fraction within ±3:', within_3)

Fraction within ±3: 0.868


**9.4 A/B Testing**: Variant A converts 48/500 users; Variant B converts 62/500. Compute conversion rates and the **absolute lift** (B−A). Simulate 10,000 bootstrap samples of the lift.

In [4]:
import numpy as np
np.random.seed(0)
conversions_a = 48
users_a = 500
conversions_b = 62
users_b = 500
rate_a = conversions_a / users_a
rate_b = conversions_b / users_b
absolute_lift = rate_b - rate_a
samples = 10000
simulated_lifts = np.empty(samples)
for i in range(samples):
    sample_a = np.random.binomial(users_a, rate_a) / users_a
    sample_b = np.random.binomial(users_b, rate_b) / users_b
    simulated_lifts[i] = sample_b - sample_a
print('Conversion rate A:', rate_a)
print('Conversion rate B:', rate_b)
print('Absolute lift (B-A):', absolute_lift)
print('Bootstrap mean lift:', simulated_lifts.mean())
print('Bootstrap std of lift:', simulated_lifts.std())

Conversion rate A: 0.096
Conversion rate B: 0.124
Absolute lift (B-A): 0.027999999999999997
Bootstrap mean lift: 0.027480400000000002
Bootstrap std of lift: 0.019623180574004816


**9.5 Sports Analytics**: Given player scores over 8 games, compute moving average with window size 3 using vectorized operations (no loops). *(Generate random integers 0–30.)*

In [5]:
import numpy as np
np.random.seed(7)
scores = np.random.randint(0, 31, size=8)
window = 3
moving_avg = np.convolve(scores, np.ones(window) / window, mode='valid')
print('Scores:', scores)
print('Moving average (window=3):', moving_avg)

Scores: [15  4 25 22  3 19 23  7]
Moving average (window=3): [14.66666667 17.         16.66666667 14.66666667 15.         16.33333333]


**9.6 Image Mini‑Task**: Create a 10×10 grayscale image as a NumPy array with a white square (value 255) in the center and 0 elsewhere. Compute its mean pixel value.

In [ ]:
import numpy as np

image = np.zeros((10, 10), dtype=int)
center = 4
image[3:7, 3:7] = 255

mean_pixel_value = image.mean()

print(image)
print('Mean pixel value:', mean_pixel_value)

**9.7 Logistics**: There are 5 warehouses with inventories `[40, 55, 20, 80, 60]` and 4 stores with demands `[30, 50, 40, 35]`. Using NumPy (not loops), compute how many units remain after fulfilling demands sequentially from warehouses.